lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+    that should be watched. Upon finding a service hook it's current ActiveState
+    will be queried from systemd to determine it's initial state.
+    """
+    global _juju_unit
+    hooks_dir = Path.cwd() / "hooks"
+    _logger.info(f"Loading services from hooks in {hooks_dir}")
+
+    if not hooks_dir.exists():
+        _logger.warning(f"Hooks dir {hooks_dir} does not exist.")
+        return
+
+    watched_services = []
+    # Get service-{service}-(started|stopped) hooks defined by the charm.
+    for hook in filter(lambda p: _service_hook_regex_filter.match(p.name), hooks_dir.iterdir()):
+        match = _service_hook_regex_filter.match(hook.name)
+        watched_services.append(match.group("service"))
+
+    _logger.info(f"Services from hooks are {watched_services}")
+    if not watched_services:
+        return
+
+    bus = await MessageBus(bus_type=BusType.SYSTEM).connect()
+
+    # Loop through all the services and be sure that a new watcher is
+    # started for new ones.
+    for service in watched_services:
+        # The .service suffix is necessary and will cause lookup failures of the
+        # service unit when readying the watcher if absent from the service name.
+        service = f"{service}.service"
+        if service not in _service_states:
+            state = await _get_state(bus, service)
+            _logger.debug(f"Adding service '{service}' with initial state: {state}")
+            _service_states[service] = state
+
+
+def _load_services(loop: asyncio.AbstractEventLoop) -> None:  # pragma: no cover
+    """Load services synchronously using _async_load_services.
+
+    This is a synchronous form of the _load_services method. This is called from a
+    signal handler which cannot take coroutines, thus this method will schedule a
+    task to run in the current running loop.
+
+    Args:
+        loop: Asynchronous event loop from main thread.
+    """
+    loop.call_soon(_async_load_services)
+
+
+async def _juju_systemd_notices_daemon() -> None:
+    """Start Juju systemd notices daemon.
+
+    This start call will set up the notices service to listen for events.
+    It connects to the system message bus and registers for signals/events on the
+    org.freedesktop.systemd1.Unit object looking for any PropertyChanged events.
+    This method additionally sets up signal handlers for various signals to either
+    terminate the process or reload the configuration from the hooks directory.
+    """
+    stop_event = asyncio.Event()
+    loop = asyncio.get_event_loop()
+    loop.add_signal_handler(signal.SIGINT, stop_event.set)
+    loop.add_signal_handler(signal.SIGTERM, stop_event.set)
+    # The event loop must be early bound to the lambda, otherwise the event loop
+    # will not exist within the lambda when the SIGHUP signal is received
+    # by the running notices daemon.
+    loop.add_signal_handler(  # pragma: no branch
+        signal.SIGHUP, lambda loop=loop: _load_services(loop)
```

Comment:
So just to clarify how this works: systemd sends `SIGHUP` to this process to say "got something new", and then `_load_services` queries the current state using the dbus API?

Reply:
SIGHUP is sent to daemons by systemd to reload the complete daemon configuration. When `juju-systemd-notices` receives the SIGHUP signal, this tells the daemon to reload the watched services based on what hooks are present in `hooks/` and enter their current state in `_service_states`. That way when new messages are received on the DBus message bus after the reload, the daemon reacts correctly instead of prematurely emitting an event because the daemon forgot the previous state of the service that it is watching.

The systemd manpage goes into more detail about signals that can be sent to services: https://manpages.ubuntu.com/manpages/jammy/en/man1/systemd.1.html#signals